### PR TITLE
[WIP] Setup LLVM automatically for LLVM backend (Mac OS/Linux)

### PR DESCRIPTION
**Pull request by d0iasm**
Setup LLVM automatically for LLVM backend. It now supports Mac OS and Linux.

# Description
Port the code from https://github.com/wasmerio/wasmer/pull/353. I created a new PR because the code in the master branch has changed a lot since the previous PR made.

The automatic installing is enabled the following conditions:
- The system LLVM doesn't exist.
- Your environment is Mac OS or Linux.
- Internet connection exists.
- The flag to disable installing isn't set.
   - `LLVM_SYS_80_DISABLE_INSTALL=1 cargo run` can disable to install LLVM.

### CODE REVIEW

**Comment by d0iasm** [comment]
bors try

**Comment by losfair** [comment]
I think it would be useful to provide an option to disable (or enable) the automatic download feature, possibly via an environmental variable. There isn't always a valid Internet connection on the build machine.

File: lib/llvm-backend/build.rs [hunk 1]
```diff
@@ -262,6 +262,46 @@ fn is_llvm_debug() -> bool {
...
+fn get_llvm_target_name() -> String {
+    let name = if cfg!(all(target_os = "macos", target_arch = "x86_64")) {
+        "x86_64-apple-darwin"
+    } else if cfg!(all(target_os = "linux", target_arch = "x86_64")) {
+        "x86_64-linux-gnu-ubuntu-16.04"
+    } else {
+        panic!("Unsupport host for installing llvm")
+    };
+
+    format!("clang+llvm-8.0.0-{}", name)
```
Comment (nlewycky) [inline lib/llvm-backend/build.rs]: Please update to 8.0.1?

File: lib/llvm-backend/build.rs [hunk 2]
```diff
@@ -262,6 +262,46 @@ fn is_llvm_debug() -> bool {
     llvm_config("--build-mode").contains("Debug")
 }
 
+fn get_llvm_target_name() -> String {
+    let name = if cfg!(all(target_os = "macos", target_arch = "x86_64")) {
+        "x86_64-apple-darwin"
+    } else if cfg!(all(target_os = "linux", target_arch = "x86_64")) {
+        "x86_64-linux-gnu-ubuntu-16.04"
+    } else {
+        panic!("Unsupport host for installing llvm")
```
Comment (nlewycky) [inline lib/llvm-backend/build.rs]:
```suggestion
        panic!("Unsupported host for installing llvm")
```

File: lib/llvm-backend/build.rs [hunk 3]
```diff
@@ -262,6 +262,46 @@ fn is_llvm_debug() -> bool {
...
+    );
+
+    if llvm_path.exists() {
+        return;
+    }
+
+    let url = find_llvm_url();
+
+    let resp = reqwest::get(&url).expect("Failed to connect to the llvm server");
+    let resp = lzma::LzmaReader::new_decompressor(resp).expect("Failed to initialize decompressor");
```
Comment (nlewycky) [inline lib/llvm-backend/build.rs]: Please have a list of expected sha256sums and verify that the download matches before doing anything else with the file.

File: lib/llvm-backend/build.rs [hunk 4]
```diff
@@ -274,6 +314,8 @@ fn main() {
     println!("cargo:rerun-if-env-changed={}", &*ENV_USE_DEBUG_MSVCRT);
     println!("cargo:rerun-if-env-changed={}", &*ENV_FORCE_FFI);
 
+    install_llvm();
```
Comment (nlewycky) [inline lib/llvm-backend/build.rs]: What happens if LLVM is already downloaded? What happens if there's a suitable LLVM 8.0 installation on the system, but not downloaded by us?

File: lib/llvm-backend/build.rs (hunk 1 above)
Comment (d0iasm) [reply lib/llvm-backend/build.rs]: I'm using pre-build binaries but it doesn't seem to exist 8.0.1 for Mac and Ubuntu 16.0.4 in the download page (http://releases.llvm.org/download.html).

File: lib/llvm-backend/build.rs (hunk 4 above)
Comment (d0iasm) [reply lib/llvm-backend/build.rs]: I added the if-condition to avoid calling `install_llvm()` when LLVM is already located on the system.

**Comment by d0iasm** [comment]
@losfair losfair
Thank you for your idea. It would also useful to check an Internet connection by code, and it won't download LLVM binary when the Internet connection doesn't exist whatever an environment variable.

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by d0iasm** [comment]
bors try

**Comment by syrusakbary** [comment]
@d0iasm it looks great.
To test that it works properly, can you remove all the lines with the following in the CI config?

```
      - template: .azure/install-llvm.yml
```

https://github.com/wasmerio/wasmer/pull/1023/files#diff-fec826feae04e51c0d94076385408bdcR117

**Comment by d0iasm** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try-

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try-

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
We might need this in the future, but we can close the PR for now

**Pull request closed without merging**